File: templates/problems/06_linked_list/10_merge_k_lists/merge_k_lists_test.go
package merge_k_lists

import (
	"reflect"
	"testing"
)

func TestMergeKLists(t *testing.T) {
	tests := []struct {
		name  string
		lists [][]int
		want  []int
	}{
		{"nil input", nil, nil},
		{"empty lists", [][]int{{}, {}, {}}, nil},
		{"single list", [][]int{{1, 2, 3}}, []int{1, 2, 3}},
		{"two lists", [][]int{{1, 4, 5}, {1, 3, 4}}, []int{1, 1, 3, 4, 4, 5}},
		{"three lists", [][]int{{1, 4, 5}, {1, 3, 4}, {2, 6}}, []int{1, 1, 2, 3, 4, 4, 5, 6}},
		{"with empty list", [][]int{{1, 2}, {}, {3, 4}}, []int{1, 2, 3, 4}},
		{"all single elements", [][]int{{5}, {1}, {3}}, []int{1, 3, 5}},
		{"duplicates across lists", [][]int{{1, 1}, {1, 1}}, []int{1, 1, 1, 1}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			var lists []*ListNode
			for _, vals := range tt.lists {
				lists = append(lists, buildList(vals))
			}
			got := listToSlice(mergeKLists(lists))
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("mergeKLists() = %v, want %v", got, tt.want)
			}
		})
	}
}

File: templates/problems/06_linked_list/10_merge_k_lists/typescript/merge_k_lists.test.ts
import { describe, it, expect } from "vitest";
import { buildList, listToSlice, ListNode } from "./helpers";
import { mergeKLists } from "./merge_k_lists";

function build(listsVals: number[][]): (ListNode | null)[] {
    return listsVals.map((vals) => buildList(vals));
}

describe("mergeKLists", () => {
    it("nil input", () => {
        expect(listToSlice(mergeKLists([]))).toEqual([]);
    });

    it("empty lists", () => {
        expect(listToSlice(mergeKLists(build([[], [], []])))).toEqual([]);
    });

    it("single list", () => {
        expect(listToSlice(mergeKLists(build([[1, 2, 3]])))).toEqual([1, 2, 3]);
    });

    it("two lists", () => {
        expect(listToSlice(mergeKLists(build([[1, 4, 5], [1, 3, 4]])))).toEqual([1, 1, 3, 4, 4, 5]);
    });

    it("three lists", () => {
        expect(listToSlice(mergeKLists(build([[1, 4, 5], [1, 3, 4], [2, 6]])))).toEqual([1, 1, 2, 3, 4, 4, 5, 6]);
    });

    it("with empty list", () => {
        expect(listToSlice(mergeKLists(build([[1, 2], [], [3, 4]])))).toEqual([1, 2, 3, 4]);
    });

    it("all single elements", () => {
        expect(listToSlice(mergeKLists(build([[5], [1], [3]])))).toEqual([1, 3, 5]);
    });

    it("duplicates across lists", () => {
        expect(listToSlice(mergeKLists(build([[1, 1], [1, 1]])))).toEqual([1, 1, 1, 1]);
    });
});
